server/utils/conditionsProvider.ts

```diff
@@ -35,3 +37,156 @@ export function getGroupedAdditionalConditions(): Record<string, unknown>[] {
   })
   return Array.from(map, ([category, conditions]) => ({ category, conditions }))
 }
+
+/**
+ * Accepts a list of additional conditions and data items for a licence and expands the
+ * placeholders with their matching data to produce a list of formatted additional conditions.
+ * @param conditions
+ */
+export function expandAdditionalConditions(conditions: AdditionalCondition[]): string[] {
+  const expandedConditions: string[] = []
+
+  conditions?.forEach(condition => {
+    const configCondition = getAdditionalConditionByCode(condition.code)
+    if (configCondition?.requiresInput) {
+      const placeholders = getPlaceholderNames(configCondition.tpl as string)
+      const inputConfig = configCondition.inputs as unknown as Record<string, unknown>[]
+      let thisConditionText = configCondition.tpl as string
+
+      for (let i = 0; i < placeholders.length; i += 1) {
+        // Remove the {} from the placeholders
+        const ph = placeholders[i].substr(1, placeholders[i].length - 2)
+
+        // Get the matching condition data by placeholder name === field name
+        const matchingDataItems = getMatchingDataItems(ph.trim(), condition.data)
+
+        if (matchingDataItems.length === 0) {
+          // Unmatched placeholder
+          thisConditionText = removeNamedPlaceholder(ph, thisConditionText)
+        } else if (matchingDataItems.length === 1) {
+          // Single matching value
+          const rules = inputConfig.find(item => item.name === ph)
+          const { value } = matchingDataItems[0]
+          thisConditionText = replacePlaceholderWithValue(
+            ph,
+            thisConditionText,
+            adjustCase(rules.case as string, value)
+          )
+        } else {
+          // List of values for this placeholder
+          const rules = inputConfig.find(item => item.name === ph)
+          const value = produceValueAsFormattedList(matchingDataItems)
+          thisConditionText = replacePlaceholderWithValue(
+            ph,
+            thisConditionText,
+            adjustCase(rules.case as string, value)
+          )
+        }
+      }
+      // Remove any that remain
+      thisConditionText = removeAllPlaceholders(thisConditionText)
+      expandedConditions.push(thisConditionText)
+    } else {
+      // No input was required - keep the condition text verbatim to print on the licence
+      expandedConditions.push(condition.text)
+    }
+  })
+
+  return expandedConditions
+}
+
+/**
+ * Formats a list of data items according to rules (add commas, and)
+ * @param matchingDataItems
+ */
+export const produceValueAsFormattedList = (matchingDataItems: AdditionalConditionData[]): string => {
+  let value = ''
+  let valueCounter = 1
+  for (let i = 0; i < matchingDataItems.length; i += 1) {
```

Comment:
matchingDataItems.forEach(item => {})

Reply:
Its the same thing here - I had it like this originally, but it never went into the forEach loop, despite it having values.